### PR TITLE
many: add support for session dbus services via snaps

### PR DESCRIPTION
**Pull request by mvo5**
This adds support for dbus activation for snaps. This version is session bus only but it will be relatively easy to extend for the system bus. It implements the proposal from Jamie in https://github.com/snapcore/snapd/pull/2592#issuecomment-272284894 and the snap.yaml for an application that implements io.snapcraft.SnapDbusService will look something like this (`service: true` is the key here):
```
name: dbus-service
slots:
  dbus-service-slot:
    interface: dbus
    bus: session
    name: io.snapcraft.SnapDbusService
    service: true
apps:
  dbus-service:
    command: bin/dbus-service
    plugs:
      - network
      - network-bind
    slots:
      - dbus-service-slot
```

LP: #1648990

### CODE REVIEW

**Comment by jdstrand** [comment]
As written, if snapd install finds meta/gui/foo.service, it takes that and creates a session service file in /var/lib/snapd/desktop/dbus-1/services and allows the desktop file to set DBusActivatable. This is simple and it works to start the session service in devmode, but is lacking security policy updates that would allow it to run in strict mode. Also, looking at what can be declared in service files, I think that we can have snapd generate them and expose the options via yaml. The dbus specification (https://dbus.freedesktop.org/doc/dbus-specification.html) is not very clear on service files but bus/desktop-file.h from the dbus sources is. These are the supported directives:

- Name=
- Exec=
- User=
- SystemdService=

We want to use 'Name=' and 'Exec='. We can ignore 'User=' for now since session services don't use it (they are expected to run as the same user as the user's session) and system services are started as root via systemd (when we support per-snap users, we'll adjust the systemd service file for services, not a dbus serevice service file). We can ignore SystemdService= since it is used by dbus-daemon to tell systemd to start the service on its behalf. We will have dbus-daemon starting session services directly and systemd starting system services directly, so this isn't needed. It is important to remember that on series 16, we have systemd user sessions, but this isn't available in 14.04, so we should have dbus-daemon start session services itself rather than systemd. 

I propose:
- move relevant parts of wrappers/service.go into interfaces/dbus/dbus.go (ie, the DBus security backend)
- have `AddSnapDBusSessionServiceFile(snapInfo, name)` create a service file:
```
[D-Bus Service]
Name=<name passed into AddSnapDBusSessionServiceFile>
Exec=<command from snap.yaml>
```
- keep DBusActivatable as something that is allowable in the desktop file
- remove 'daemon: dbus' from snap info since it is at this point vestigial code

With the above, we have the pieces in place to create session service files, but no way to declare they should be used or how to generate the security policy. I think we can look at the dbus interface and add a slot side attribute. We currently have:

```
name: (slot): well-known DBus connection name (eg, org.foo.bar)
bus (slot): DBus bus to use (ie, session or system)
```

We add:
```
service (slot): true|false. When true, the command is started as a DBus service. Defaults
to false (ie, if command uses 'daemon', started via systemd without using 'Type=dbus',
otherwise started by the user)
```

Then the dbus interface can be adjusted such that:
- if service is not specified, no changes to the current implementation
- if service == True and bus == 'session', call AddSnapDBusSessionServiceFile(s, 'session', name)
- if service == True and bus == 'system', when the systemd unit file is generated, use `BusName=<name>` and `Type=dbus`

With this, the `dbus` interface allows people to create session services started by dbus-daemon (that may be dbus activatable if the desktop file has DBusActivatable=true) and to prefer 'Type=dbus' for system services started by systemd. Future interfaces could leverage AddSnapDBusSessionServiceFile() as well for creating session services.

CC: @vosst, @niemeyer, @tyhicks, @ejratl

**Comment by mvo5** [comment]
@jdstrand Thanks for the excellent suggestions. I changed the code as you suggested. 

Some limitations with this iteration:
- no dbus system service file generation yet
- the implementation is hackish (setupDbusservices), I think once #2613 lands this will become less ugly 
- unit tests missing

**Comment by mvo5** [comment]
Fwiw, the 14.04 failure can be ignored, the packaging branch needs an update and then it should work.

**Comment by jhenstridge** [comment]
Would it make sense to also generate a `SystemdService` key in the generated D-Bus .service file?

If we're assuming systemd for system and user init, then this would reduce the number of code paths for launching confined daemons on the system.

File: interfaces/dbus/backend.go [hunk 1]
```diff
@@ -132,6 +156,68 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
 
+func (b *Backend) setupBusservices(snapInfo *snap.Info, snippets map[string][][]byte) error {
+	snapName := snapInfo.Name()
+
+	content := map[string]*osutil.FileState{}
+	for _, appInfo := range snapInfo.Apps {
+		securityTag := appInfo.SecurityTag()
+
+		// FIXME: layer violation, we really should do that in the
+		//        dbus interface but without PR #2613 its really tricky
```
Comment (niemeyer) [inline interfaces/dbus/backend.go]: #2613 is now in. Seems okay to fix that in a follow up, though, assuming everything else is golden.

File: interfaces/dbus/backend.go [hunk 2]
```diff
@@ -132,6 +156,68 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+				Content: buffer.Bytes(),
+				Mode:    0644,
+			}
+		}
+	}
+
+	glob := fmt.Sprintf("%s.service", interfaces.SecurityTagGlob(snapName))
+	dir := dirs.SnapDBusSessionServicesFilesDir
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for DBus service files %q: %s", dir, err)
```
Comment (niemeyer) [inline interfaces/dbus/backend.go]: The %q/dir is probably redundant here. The error from MkdirAll will include the path name as well.

File: interfaces/dbus/backend.go [hunk 3]
```diff
@@ -132,6 +156,68 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+		//        dbus interface but without PR #2613 its really tricky
+		for _, slot := range appInfo.Slots {
+			if slot.Interface != "dbus" {
+				continue
+			}
+			bus, ok := slot.Attrs["bus"].(string)
+			if !ok {
+				continue
+			}
+			if bus != "session" {
```
Comment (niemeyer) [inline interfaces/dbus/backend.go]: Shouldn't we allow "system" as well?

File: data/dbus/snapd.conf [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<busconfig>
+ <servicedir>/var/lib/snapd/dbus-1/services/</servicedir>
```
Comment (niemeyer) [inline data/dbus/snapd.conf]: Can we please use "dbus/services"? That "dbus-1" convention always feels awkward to me, and we don't have many reasons to reproduce it here I think. The compatibility issues and their solution will be the same as for every other service we hold data for.

File: wrappers/desktop_test.go [hunk 5]
```diff
@@ -300,8 +300,14 @@ func (s *sanitizeDesktopFileSuite) TestLangLang(c *C) {
 		// bad ones
 		{"Name[foo=bar", false},
 		{"Icon[xx]=bar", false},
+		// dbus related
+		{"Activactable=true", false},
```
Comment (niemeyer) [inline wrappers/desktop_test.go]: Is the typo intended here?

File: interfaces/dbus/backend.go [hunk 6]
```diff
@@ -132,6 +156,68 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+				continue
+			}
+			name, ok := slot.Attrs["name"].(string)
+			if !ok {
+				continue
+			}
+			isService := slot.Attrs["service"].(bool)
```
Comment (niemeyer) [inline interfaces/dbus/backend.go]: Even without #2613, can we please at least validate that setting there, and cross-link it here (mentioning this is being used with this purpose).

File: data/dbus/snapd.conf (hunk 4 above)
Comment (niemeyer) [inline data/dbus/snapd.conf]: This will need fixes elsewhere too.

File: wrappers/desktop.go [hunk 7]
```diff
@@ -100,13 +100,14 @@ var isValidDesktopFileLine = regexp.MustCompile(strings.Join([]string{
 	// unity extension
 	"^X-Ayatana-Desktop-Shortcuts=",
 	"^TargetEnvironment=",
+	// dbus
+	"^DBusActivatable=(true|false)",
 }, "|")).MatchString
 
 // rewriteExecLine rewrites a "Exec=" line to use the wrapper path for snap application.
-func rewriteExecLine(s *snap.Info, desktopFile, line string) (string, error) {
+func rewriteExecLine(s *snap.Info, desktopFile, line, env string) (string, error) {
```
Comment (mvo5) [inline wrappers/desktop.go]: That was from a previous iteration, this is fixed now.

File: interfaces/dbus/backend.go [hunk 8]
```diff
@@ -50,14 +50,27 @@ func (b *Backend) Name() interfaces.SecuritySystem {
...
 	snippets, err := repo.SecuritySnippetsForSnap(snapInfo.Name(), interfaces.SecurityDBus)
 	if err != nil {
-		return fmt.Errorf("cannot obtain DBus security snippets for snap %q: %s", snapName, err)
+		return fmt.Errorf("cannot obtain DBus security snippets for snap %q: %s", snapInfo.Name(), err)
+	}
+
+	if err := b.setupBusconfig(snapInfo, snippets); err != nil {
+		return err
 	}
+	if err := b.setupBusservices(snapInfo, snippets); err != nil {
```
Comment (mvo5) [inline interfaces/dbus/backend.go]: Thanks, fixed.

File: interfaces/dbus/backend.go (hunk 1 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: Thanks, +1 for a followup

File: interfaces/dbus/backend.go (hunk 3 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: I think we want that in the longer run, my main focus for this PR was to unblock people who need dbus activated services (I think unity and kde has some of those).

File: interfaces/dbus/backend.go (hunk 6 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: I added checks that its really bool and also added a comment why we need this. Should I do more (I'm not sure if that addresses the  cross-link comment).

File: interfaces/dbus/backend.go (hunk 2 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: thanks, removed

File: data/dbus/snapd.conf (hunk 4 above)
Comment (mvo5) [inline data/dbus/snapd.conf]: updated.

File: wrappers/desktop_test.go (hunk 5 above)
Comment (mvo5) [inline wrappers/desktop_test.go]: ups, I fixed that typo, the test was focused on the missing "DBus" in front, not so much on the typo :)

**Comment by jdstrand** [comment]
@jhenstridge re 'SystemdService key', I covered this in:  https://github.com/snapcore/snapd/pull/2592#issuecomment-272284894

File: interfaces/dbus/backend.go [hunk 9]
```diff
@@ -50,14 +50,27 @@ func (b *Backend) Name() interfaces.SecuritySystem {
...
+		return err
 	}
+	if err := b.setupBusServ(snapInfo, snippets); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (b *Backend) setupBusConf(snapInfo *snap.Info, snippets map[string][][]byte) error {
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: Can you add a comment for what this function does? Eg:
```
// Create the DBus busconfig policy for the snap
```

File: interfaces/dbus/backend.go [hunk 10]
```diff
@@ -77,6 +90,16 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
...
+	if err := b.removeBusConf(snapName); err != nil {
+		return err
+	}
+	if err := b.removeBusServ(snapName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Backend) removeBusConf(snapName string) error {
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: Can you add a comment above this function. Eg
```
// Remove the DBus busconfig policy for the snap
```

File: interfaces/dbus/backend.go [hunk 11]
```diff
@@ -85,9 +108,10 @@ func (b *Backend) Remove(snapName string) error {
 	return nil
 }
 
-// combineSnippets combines security snippets collected from all the interfaces
-// affecting a given snap into a content map applicable to EnsureDirState.
-func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]byte) (content map[string]*osutil.FileState, err error) {
+// combineSnippetsBusConf combines security snippets collected from
+// all the interfaces affecting a given snap into a content map
+// applicable to EnsureDirState.
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: Can you rephrase this to be:
```
// combineSnippetsBusConf combines security snippets for busconfig
// policy collected from all the interfaces affecting a given snap into a
// content map applicable to EnsureDirState.
```

File: interfaces/dbus/backend.go [hunk 12]
```diff
@@ -111,13 +135,13 @@ func (b *Backend) combineSnippets(snapInfo *snap.Info, snippets map[string][][]b
...
 
-		addContent(securityTag, hookSnippets, content)
+		addContentBusConf(securityTag, hookSnippets, content)
 	}
 
 	return content, nil
 }
 
-func addContent(securityTag string, executableSnippets [][]byte, content map[string]*osutil.FileState) {
+func addContentBusConf(securityTag string, executableSnippets [][]byte, content map[string]*osutil.FileState) {
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: Can you add a comment above this. Eg:
```
// addContentBusConf creates/updates the xml busconfig policy for the snap
```

File: interfaces/dbus/backend.go (hunk 3 above)
Comment (jdstrand) [inline interfaces/dbus/backend.go]: If you aren't going to support system now, can you add a TODO. Eg:
```
// TODO: we can eventually support 'system' by:
// 1. creating the SnapDBusSystemServicesFilesDir directory
// 2. writing the service file to SnapDBusSystemServicesFilesDir
//    when 'daemon' is set to 'dbus' (see validate.go)
// 3. add 'Type=dbus' and 'BusName=slot.Attrs["name"].(string)'
//    to the systemd unit when 'slot.Attrs["service"].(bool) == True'
//    and 'daemon' is set to 'dbus'

File: interfaces/dbus/backend.go [hunk 13]
```diff
@@ -132,6 +156,70 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+			}
+			if bus != "session" {
+				continue
+			}
+			name, ok := slot.Attrs["name"].(string)
+			if !ok {
+				continue
+			}
+			// we check if its a service here so that we know
+			// if a dbus service file needs to get generated.
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: s/to get/to be/

File: interfaces/dbus/backend.go [hunk 14]
```diff
@@ -132,6 +156,70 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+			isService, ok := slot.Attrs["service"].(bool)
+			if !ok || !isService {
+				continue
+			}
+
+			var buffer bytes.Buffer
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: Can you add a comment here. Eg:
```
// We set only 'Name' and 'Exec' for now. We may add 'User' for 'system'
// services when we support per-snap users. Don't specify 'SystemdService'
// and just let dbus-daemon launch the service since 'SystemdService' is only
// used by dbus-daemon to tell systemd to launch the service and systemd
// user sessions aren't available everywhere yet.

File: snap/validate.go [hunk 15]
```diff
@@ -134,7 +134,7 @@ var validAppName = regexp.MustCompile("^[a-zA-Z0-9](?:-?[a-zA-Z0-9])*$")
 // ValidateApp verifies the content in the app info.
 func ValidateApp(app *AppInfo) error {
 	switch app.Daemon {
-	case "", "simple", "forking", "oneshot", "dbus", "notify":
+	case "", "simple", "forking", "oneshot", "notify":
```
Comment (jdstrand) [inline snap/validate.go]: Please add:
```
// TODO: add 'dbus' when we support 'service: true' in the dbus interface
```

File: interfaces/dbus/backend.go (hunk 6 above)
Comment (niemeyer) [inline interfaces/dbus/backend.go]: Should be enough, thanks.

**Comment by niemeyer** [comment]
Test failures are related:
```
2017/02/15 09:07:51 Error restoring linode:ubuntu-14.04-64:tests/main/dbus-activated : 
-----
+ killall dbus-daemon
+ apt autoremove -y dbus-x11
E: Command line option 'y' [from -y] is not known.
-----

File: interfaces/dbus/backend.go (hunk 10 above)
Comment (jdstrand) [inline interfaces/dbus/backend.go]: @mvo5 - nitpick, this comment still didn't make it in.

File: interfaces/dbus/backend.go [hunk 16]
```diff
@@ -132,6 +158,88 @@ func addContent(securityTag string, executableSnippets [][]byte, content map[str
...
+		return fmt.Errorf("cannot create directory for DBus service files: %s", err)
+	}
+	_, _, err := osutil.EnsureDirState(dir, glob, content)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize DBus service files for snap %q: %s", snapName, err)
+	}
+	return nil
+}
+
+// Remove the DBus busconfig policy for the snap
```
Comment (jdstrand) [inline interfaces/dbus/backend.go]: This comment should be:
```
// Remove the DBus service file for the snap
```

File: data/dbus/snapd.conf [hunk 17]
```diff
@@ -0,0 +1,3 @@
+<busconfig>
```
Comment (mvo5) [inline data/dbus/snapd.conf]: This will not work with re-exec, i.e. this config file needs to be in the regular deb package right now and if snapd just re-execs it won't be there. So we probably need to add code that writes it if its missing. Which has the downside that it will also mean that we leave files behind when snapd gets removed.

Comment (niemeyer) [inline data/dbus/snapd.conf]: That's already the case with pretty much all of the security backends, and we actually already have a dbus backend that is tasked precisely with writing out that sort of file, and we ought to be removing these already when the package goes away (are we?).

**Comment by mvo5** [comment]
Closing for now as this needs further work

**Comment by mvo5** [comment]
Reopening @morphis was interessted.

File: data/dbus/snapd.conf (hunk 17 above)
Comment (mvo5) [inline data/dbus/snapd.conf]: This is now auto-generated if missing

File: interfaces/dbus/backend.go [hunk 18]
```diff
@@ -86,9 +166,18 @@ func (b *Backend) Remove(snapName string) error {
...
+	glob := fmt.Sprintf("%s.service", interfaces.SecurityTagGlob(snapName))
+	_, _, err := osutil.EnsureDirState(dirs.SnapBusPolicyDir, glob, nil)
+	if err != nil {
+		return fmt.Errorf("cannot synchronize DBus service files for snap %q: %s", snapName, err)
+	}
+	return nil
+}
+
 // deriveContent combines security snippets collected from all the interfaces
 // affecting a given snap into a content map applicable to EnsureDirState.
```
Comment (mvo5) [inline interfaces/dbus/backend.go]: Thanks, added

File: dirs/dirs.go [hunk 19]
```diff
@@ -138,6 +140,8 @@ func SetRootDir(rootdir string) {
 	SnapMetaDir = filepath.Join(rootdir, snappyDir, "meta")
 	SnapBlobDir = filepath.Join(rootdir, snappyDir, "snaps")
 	SnapDesktopFilesDir = filepath.Join(rootdir, snappyDir, "desktop", "applications")
+	SnapDBusSessionServicesFilesDir = filepath.Join(rootdir, snappyDir, "dbus/services")
```
Comment (mvo5) [inline dirs/dirs.go]: Good point, added

File: interfaces/builtin/dbus.go [hunk 20]
```diff
@@ -316,13 +326,30 @@ func (iface *DbusInterface) DBusPermanentSlot(spec *dbus.Specification, slot *in
 	}
 
 	// only system services need bus policy
```
Comment (mvo5) [inline interfaces/builtin/dbus.go]: Nice, updated

File: interfaces/builtin/dbus.go [hunk 21]
```diff
@@ -316,13 +326,30 @@ func (iface *DbusInterface) DBusPermanentSlot(spec *dbus.Specification, slot *in
...
-	if bus != "system" {
-		return nil
+	if bus == "system" {
+		old := "###DBUS_NAME###"
+		new := name
+		spec.AddSnippet(strings.Replace(dbusPermanentSlotDBus, old, new, -1))
+	}
+
+	// dbus activation currently only supported by the session bus
```
Comment (mvo5) [inline interfaces/builtin/dbus.go]: :+1:

File: interfaces/dbus/backend.go [hunk 22]
```diff
@@ -46,6 +50,29 @@ func (b *Backend) Name() interfaces.SecuritySystem {
 	return "dbus"
 }
 
+// setupHostDBusSessionConf will ensure that we have a dbus configuration
+// that points to /var/lib/snapd/dbus/services. This is needed for systems
+// that re-exec snapd and do not have this configuration as part of the
+// packaged snapd.
+func setupHostDBusSessionConf() error {
+	dbusSessionConf := filepath.Join(dirs.SnapSessionBusPolicyDir, "snapd.conf")
+	if osutil.FileExists(dbusSessionConf) {
```
Comment (zyga) [inline interfaces/dbus/backend.go]: Could we use osutil.EnsureDirState for this? This will allow us to just have a simple call here that ensures the _correct_ content is in that file.

File: interfaces/dbus/backend.go [hunk 23]
```diff
@@ -74,10 +118,46 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	return nil
 }
 
+func (b *Backend) setupBusActivatedSessionServ(snapInfo *snap.Info, spec interfaces.Specification) error {
+	snapName := snapInfo.Name()
+
+	content := map[string]*osutil.FileState{}
+	for securityTag, serviceContent := range spec.(*Specification).SessionServices() {
+		fname := fmt.Sprintf("%s.service", securityTag)
```
Comment (zyga) [inline interfaces/dbus/backend.go]: Are all security tag valid systemd unit file names?

File: interfaces/dbus/backend.go [hunk 24]
```diff
@@ -74,10 +118,46 @@ func (b *Backend) Setup(snapInfo *snap.Info, opts interfaces.ConfinementOptions,
 	return nil
 }
 
+func (b *Backend) setupBusActivatedSessionServ(snapInfo *snap.Info, spec interfaces.Specification) error {
+	snapName := snapInfo.Name()
+
+	content := map[string]*osutil.FileState{}
```
Comment (zyga) [inline interfaces/dbus/backend.go]: Could you please move this to a helper similar to `deriveContentBusConf`?

File: interfaces/dbus/spec.go [hunk 25]
```diff
@@ -21,16 +21,50 @@ package dbus
...
+		// 'User' for 'system' services when we support
+		// per-snap users. Don't specify 'SystemdService' and
+		// just let dbus-daemon launch the service since
+		// 'SystemdService' is only used by dbus-daemon to
+		// tell systemd to launch the service and systemd user
+		// sessions aren't available everywhere yet.
+		spec.sessionServices[appInfo.SecurityTag()] = fmt.Sprintf(`[D-BUS Service]
+Name=%s
+Exec=%s
+`, name, appInfo.LauncherCommand())
```
Comment (zyga) [inline interfaces/dbus/spec.go]: Note that we have `systemd.Service` and the `systemd` backend as well. We could perhaps reuse some of those here.

File: interfaces/dbus/backend.go (hunk 23 above)
Comment (morphis) [inline interfaces/dbus/backend.go]: If not we can pipe the security tag through systemd-escape, which is maybe safe in any case when we generate a systemd unit name.

File: interfaces/builtin/dbus.go [hunk 26]
```diff
@@ -398,7 +425,17 @@ func (iface *DbusInterface) SanitizeSlot(slot *interfaces.Slot) error {
 	}
 
 	_, _, err := iface.getAttribs(slot.Attrs)
-	return err
+	if err != nil {
+		return err
+	}
+
+	// FIXME: also check that the dbus name is not already taken
+	//        by an existing snap
```
Comment (morphis) [inline interfaces/builtin/dbus.go]: How can we reliable implement this? Is there any way to query dbus for this as I think examining available configuration files isn't enough. org.freedesktop.DBus.ListActivatableNames looks like a good candidate for this. I am adding godbus to the snapd source tree already with https://github.com/snapcore/snapd/pull/3260 so would be fine to add it with this PR too.

Comment (jdstrand) [inline interfaces/builtin/dbus.go]: The FIXME is for if a snap has taken it, not if anything has taken it. snapd has all the information for this: specifically all the snaps on the system with connected dbus interfaces and the attributes of those interfaces.

I don't think we want to query DBus only if we want to make this generalizable beyond activatable names because it would be useful if a non-activatable names were also checked. Right now, those snaps simply fail to start, but we could do something better if we were smarter with existing connections. snapd could additionally look at ListActivatableNames for the activated services to check for non-snap activated service collisions.

**Comment by pedronis** [comment]
see discussion in #3323 whether we need to tweak wrappers.IsService here

File: interfaces/dbus/spec.go (hunk 25 above)
Comment (mvo5) [inline interfaces/dbus/spec.go]: In this particular case we write a dbus service file which is different from a systemd unit file.

File: interfaces/dbus/backend.go (hunk 24 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: I reworked the code to do that now.

File: interfaces/dbus/backend.go (hunk 22 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: This is done now.

File: interfaces/dbus/backend.go (hunk 23 above)
Comment (mvo5) [inline interfaces/dbus/backend.go]: The files written here are dbus service files. The spec https://dbus.freedesktop.org/doc/dbus-specification.html has no restrictions on the name AFAICS.

**Comment by mvo5** [comment]
@pedronis I think it does make sense to make wrapper.IsService() return true for dbus activated services. This way we also avoid writing the /snap/bin/svc binary and avoid polluting the namespace of PATH with those.

**Comment by mvo5** [comment]
Closing for now as this needs more work but there is no time available right now, I will reopen once I get the chance to work on it again.

**Comment by PixsaOJ** [comment]
@mvo5 Did you get a chance to work on it again? 😃

**Comment by jhenstridge** [comment]
@PixsaOJ: D-Bus activation support is available in snapd. Support for session bus services is currently gated behind an experimental feature flag. You can see some more details here:

https://forum.snapcraft.io/t/enabling-user-daemons-and-d-bus-activation/22318?u=jamesh
